File: csharp/Prescribe.Core/Frontend/Ast.cs
namespace Prescribe.Core.Frontend;

public readonly record struct SourceLoc(int Line, int Column);

public abstract record AstNode(SourceLoc Loc);

public sealed record ProgramNode(string Name, BlockNode Block, SourceLoc Loc) : AstNode(Loc);

public sealed record BlockNode(IReadOnlyList<DeclarationNode> Declarations, IReadOnlyList<StatementNode> Statements, SourceLoc Loc) : AstNode(Loc);

public abstract record DeclarationNode(SourceLoc Loc) : AstNode(Loc);

public enum AccessModifier
{
    Public,
    Private
}

public interface IClassMember { }

public sealed record VarDeclNode(string Name, TypeNode TypeSpec, SourceLoc Loc, AccessModifier? Access = null) : DeclarationNode(Loc), IClassMember;

public sealed record ConstDeclNode(string Name, ExprNode Expr, SourceLoc Loc) : DeclarationNode(Loc);

public sealed record TypeDeclNode(string Name, TypeNode TypeSpec, SourceLoc Loc) : DeclarationNode(Loc);

public enum ParamMode
{
    ByVal,
    ByRef
}

public sealed record ParamNode(string Name, ParamMode Mode, TypeNode TypeSpec, SourceLoc Loc) : AstNode(Loc);

public sealed record ProcDeclNode(string Name, IReadOnlyList<ParamNode> Params, BlockNode Block, SourceLoc Loc, AccessModifier? Access = null) : DeclarationNode(Loc), IClassMember;

public sealed record FuncDeclNode(string Name, IReadOnlyList<ParamNode> Params, TypeNode ReturnType, BlockNode Block, SourceLoc Loc, AccessModifier? Access = null) : DeclarationNode(Loc), IClassMember;

public sealed record ConstructorDeclNode(string Name, IReadOnlyList<ParamNode> Params, BlockNode Block, SourceLoc Loc, AccessModifier? Access = null) : DeclarationNode(Loc), IClassMember;

public sealed record ClassDeclNode(string Name, string? BaseName, IReadOnlyList<IClassMember> Members, SourceLoc Loc) : DeclarationNode(Loc);

public abstract record StatementNode(SourceLoc Loc) : AstNode(Loc);

public sealed record AssignStmtNode(LValueNode Target, ExprNode Expr, SourceLoc Loc) : StatementNode(Loc);

public sealed record IfStmtNode(ExprNode Condition, BlockNode ThenBlock, BlockNode? ElseBlock, SourceLoc Loc) : StatementNode(Loc);

public sealed record CaseBranchNode(IReadOnlyList<CaseLabelNode> Labels, BlockNode Block, SourceLoc Loc) : AstNode(Loc);

public abstract record CaseLabelNode(SourceLoc Loc) : AstNode(Loc);

public sealed record CaseValueNode(LiteralNode Value, SourceLoc Loc) : CaseLabelNode(Loc);

public sealed record CaseRangeNode(LiteralNode Start, LiteralNode End, SourceLoc Loc) : CaseLabelNode(Loc);

public sealed record CaseStmtNode(ExprNode Expr, IReadOnlyList<CaseBranchNode> Branches, BlockNode? OtherwiseBlock, SourceLoc Loc) : StatementNode(Loc);

public sealed record ForStmtNode(string Name, ExprNode Start, ExprNode End, ExprNode? Step, BlockNode Block, SourceLoc Loc) : StatementNode(Loc);

public sealed record WhileStmtNode(ExprNode Condition, BlockNode Block, SourceLoc Loc) : StatementNode(Loc);

public sealed record RepeatStmtNode(BlockNode Block, ExprNode Condition, SourceLoc Loc) : StatementNode(Loc);

public sealed record CallStmtNode(ProcRefNode Callee, IReadOnlyList<ExprNode> Args, SourceLoc Loc) : StatementNode(Loc);

public sealed record ReturnStmtNode(ExprNode? Expr, SourceLoc Loc) : StatementNode(Loc);

public sealed record InputStmtNode(IReadOnlyList<LValueNode> Targets, SourceLoc Loc) : StatementNode(Loc);

public sealed record OutputStmtNode(IReadOnlyList<ExprNode> Values, SourceLoc Loc) : StatementNode(Loc);

public abstract record FileStmtNode(SourceLoc Loc) : StatementNode(Loc);

public sealed record OpenFileStmtNode(string FileName, string Path, string Mode, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record CloseFileStmtNode(string FileName, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record ReadFileStmtNode(string FileName, LValueNode Target, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record WriteFileStmtNode(string FileName, ExprNode Expr, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record SeekStmtNode(string FileName, ExprNode Address, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record GetRecordStmtNode(string FileName, LValueNode Target, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record PutRecordStmtNode(string FileName, ExprNode Expr, SourceLoc Loc) : FileStmtNode(Loc);

public sealed record SuperCallStmtNode(string? MethodName, IReadOnlyList<ExprNode> Args, SourceLoc Loc) : StatementNode(Loc);

public sealed record ProcRefNode(IReadOnlyList<string> Parts, SourceLoc Loc) : AstNode(Loc);

public abstract record ExprNode(SourceLoc Loc) : AstNode(Loc);

public sealed record BinaryExprNode(string Op, ExprNode Left, ExprNode Right, SourceLoc Loc) : ExprNode(Loc);

public sealed record UnaryExprNode(string Op, ExprNode Expr, SourceLoc Loc) : ExprNode(Loc);

public sealed record LiteralNode(object Value, LiteralType LiteralType, SourceLoc Loc) : ExprNode(Loc);

public enum LiteralType
{
    Integer,
    Real,
    Boolean,
    Char,
    String,
    Date
}

public sealed record NameExprNode(string Name, SourceLoc Loc) : LValueNode(Loc);

public sealed record CallExprNode(ExprNode Callee, IReadOnlyList<ExprNode> Args, SourceLoc Loc) : ExprNode(Loc);

public sealed record IndexExprNode(ExprNode Base, IReadOnlyList<ExprNode> Indices, SourceLoc Loc) : LValueNode(Loc);

public sealed record FieldExprNode(ExprNode Base, string Field, SourceLoc Loc) : LValueNode(Loc);

public sealed record NewExprNode(string? TypeName, TypeNode? TypeSpec, IReadOnlyList<ExprNode> Args, SourceLoc Loc) : ExprNode(Loc);

public sealed record EOFExprNode(string FileName, SourceLoc Loc) : ExprNode(Loc);

public sealed record NullExprNode(SourceLoc Loc) : ExprNode(Loc);

public abstract record LValueNode(SourceLoc Loc) : ExprNode(Loc);

public sealed record DerefExprNode(ExprNode Expr, SourceLoc Loc) : LValueNode(Loc);

public abstract record TypeNode(SourceLoc Loc) : AstNode(Loc);

public sealed record BasicTypeNode(BasicTypeName Name, SourceLoc Loc) : TypeNode(Loc);

public enum BasicTypeName
{
    INTEGER,
    REAL,
    BOOLEAN,
    CHAR,
    STRING,
    DATE
}

public sealed record ArrayTypeNode(IReadOnlyList<ArrayBoundNode> Bounds, TypeNode ElementType, SourceLoc Loc) : TypeNode(Loc);

public sealed record ArrayBoundNode(int Low, int High, SourceLoc Loc) : AstNode(Loc);

public sealed record RecordTypeNode(IReadOnlyList<VarDeclNode> Fields, SourceLoc Loc) : TypeNode(Loc);

public sealed record EnumTypeNode(IReadOnlyList<string> Members, SourceLoc Loc) : TypeNode(Loc);

public sealed record SetTypeNode(string BaseName, SourceLoc Loc) : TypeNode(Loc);

public sealed record PointerTypeNode(TypeNode Target, SourceLoc Loc) : TypeNode(Loc);

public abstract record FileTypeNode(SourceLoc Loc) : TypeNode(Loc);

public sealed record TextFileTypeNode(SourceLoc Loc) : FileTypeNode(Loc);

public sealed record RandomFileTypeNode(string RecordName, SourceLoc Loc) : FileTypeNode(Loc);

public sealed record NamedTypeNode(string Name, SourceLoc Loc) : TypeNode(Loc);

File: legacy-typescript/src/frontend/ast.ts
export type SourceLoc = { line: number; column: number };

export type ProgramNode = {
  kind: "Program";
  name: string;
  block: BlockNode;
  loc: SourceLoc;
};

export type BlockNode = {
  kind: "Block";
  declarations: DeclarationNode[];
  statements: StatementNode[];
  loc: SourceLoc;
};

export type DeclarationNode =
  | VarDeclNode
  | ConstDeclNode
  | TypeDeclNode
  | ProcDeclNode
  | FuncDeclNode
  | ClassDeclNode;

export type VarDeclNode = {
  kind: "VarDecl";
  name: string;
  typeSpec: TypeNode;
  loc: SourceLoc;
  access?: "PUBLIC" | "PRIVATE";
};

export type ConstDeclNode = {
  kind: "ConstDecl";
  name: string;
  expr: ExprNode;
  loc: SourceLoc;
};

export type TypeDeclNode = {
  kind: "TypeDecl";
  name: string;
  typeSpec: TypeNode;
  loc: SourceLoc;
};

export type ParamNode = {
  name: string;
  mode: "BYVAL" | "BYREF";
  typeSpec: TypeNode;
  loc: SourceLoc;
};

export type ProcDeclNode = {
  kind: "ProcDecl";
  name: string;
  params: ParamNode[];
  block: BlockNode;
  loc: SourceLoc;
  access?: "PUBLIC" | "PRIVATE";
};

export type FuncDeclNode = {
  kind: "FuncDecl";
  name: string;
  params: ParamNode[];
  returnType: TypeNode;
  block: BlockNode;
  loc: SourceLoc;
  access?: "PUBLIC" | "PRIVATE";
};

export type ConstructorDeclNode = {
  kind: "ConstructorDecl";
  name: string;
  params: ParamNode[];
  block: BlockNode;
  loc: SourceLoc;
  access?: "PUBLIC" | "PRIVATE";
};

export type ClassDeclNode = {
  kind: "ClassDecl";
  name: string;
  baseName?: string;
  members: ClassMemberNode[];
  loc: SourceLoc;
};

export type ClassMemberNode = VarDeclNode | ProcDeclNode | FuncDeclNode | ConstructorDeclNode;

export type StatementNode =
  | AssignStmtNode
  | IfStmtNode
  | CaseStmtNode
  | ForStmtNode
  | WhileStmtNode
  | RepeatStmtNode
  | CallStmtNode
  | ReturnStmtNode
  | InputStmtNode
  | OutputStmtNode
  | FileStmtNode
  | SuperCallStmtNode;

export type AssignStmtNode = {
  kind: "AssignStmt";
  target: LValueNode;
  expr: ExprNode;
  loc: SourceLoc;
};

export type IfStmtNode = {
  kind: "IfStmt";
  condition: ExprNode;
  thenBlock: BlockNode;
  elseBlock?: BlockNode;
  loc: SourceLoc;
};

export type CaseBranchNode = {
  labels: CaseLabelNode[];
  block: BlockNode;
  loc: SourceLoc;
};

export type CaseLabelNode =
  | { kind: "CaseValue"; value: LiteralNode; loc: SourceLoc }
  | { kind: "CaseRange"; start: LiteralNode; end: LiteralNode; loc: SourceLoc };

export type CaseStmtNode = {
  kind: "CaseStmt";
  expr: ExprNode;
  branches: CaseBranchNode[];
  otherwiseBlock?: BlockNode;
  loc: SourceLoc;
};

export type ForStmtNode = {
  kind: "ForStmt";
  name: string;
  start: ExprNode;
  end: ExprNode;
  step?: ExprNode;
  block: BlockNode;
  loc: SourceLoc;
};

export type WhileStmtNode = {
  kind: "WhileStmt";
  condition: ExprNode;
  block: BlockNode;
  loc: SourceLoc;
};

export type RepeatStmtNode = {
  kind: "RepeatStmt";
  block: BlockNode;
  condition: ExprNode;
  loc: SourceLoc;
};

export type CallStmtNode = {
  kind: "CallStmt";
  callee: ProcRefNode;
  args: ExprNode[];
  loc: SourceLoc;
};

export type ReturnStmtNode = {
  kind: "ReturnStmt";
  expr?: ExprNode;
  loc: SourceLoc;
};

export type InputStmtNode = {
  kind: "InputStmt";
  targets: LValueNode[];
  loc: SourceLoc;
};

export type OutputStmtNode = {
  kind: "OutputStmt";
  values: ExprNode[];
  loc: SourceLoc;
};

export type FileStmtNode =
  | OpenFileStmtNode
  | CloseFileStmtNode
  | ReadFileStmtNode
  | WriteFileStmtNode
  | SeekStmtNode
  | GetRecordStmtNode
  | PutRecordStmtNode;

export type OpenFileStmtNode = {
  kind: "OpenFileStmt";
  fileName: string;
  path: string;
  mode: string;
  loc: SourceLoc;
};

export type CloseFileStmtNode = {
  kind: "CloseFileStmt";
  fileName: string;
  loc: SourceLoc;
};

export type ReadFileStmtNode = {
  kind: "ReadFileStmt";
  fileName: string;
  target: LValueNode;
  loc: SourceLoc;
};

export type WriteFileStmtNode = {
  kind: "WriteFileStmt";
  fileName: string;
  expr: ExprNode;
  loc: SourceLoc;
};

export type SeekStmtNode = {
  kind: "SeekStmt";
  fileName: string;
  address: ExprNode;
  loc: SourceLoc;
};

export type GetRecordStmtNode = {
  kind: "GetRecordStmt";
  fileName: string;
  target: LValueNode;
  loc: SourceLoc;
};

export type PutRecordStmtNode = {
  kind: "PutRecordStmt";
  fileName: string;
  expr: ExprNode;
  loc: SourceLoc;
};

export type SuperCallStmtNode = {
  kind: "SuperCallStmt";
  methodName?: string;
  args: ExprNode[];
  loc: SourceLoc;
};

export type ProcRefNode = {
  kind: "ProcRef";
  parts: string[];
  loc: SourceLoc;
};

export type ExprNode =
  | BinaryExprNode
  | UnaryExprNode
  | LiteralNode
  | NameExprNode
  | CallExprNode
  | IndexExprNode
  | FieldExprNode
  | NewExprNode
  | EOFExprNode
  | NullExprNode
  | DerefExprNode;

export type BinaryExprNode = {
  kind: "BinaryExpr";
  op: string;
  left: ExprNode;
  right: ExprNode;
  loc: SourceLoc;
};

export type UnaryExprNode = {
  kind: "UnaryExpr";
  op: string;
  expr: ExprNode;
  loc: SourceLoc;
};

export type LiteralNode = {
  kind: "Literal";
  value: unknown;
  literalType: "Integer" | "Real" | "Boolean" | "Char" | "String" | "Date";
  loc: SourceLoc;
};

export type NameExprNode = {
  kind: "NameExpr";
  name: string;
  loc: SourceLoc;
};

export type CallExprNode = {
  kind: "CallExpr";
  callee: ExprNode;
  args: ExprNode[];
  loc: SourceLoc;
};

export type IndexExprNode = {
  kind: "IndexExpr";
  base: ExprNode;
  indices: ExprNode[];
  loc: SourceLoc;
};

export type FieldExprNode = {
  kind: "FieldExpr";
  base: ExprNode;
  field: string;
  loc: SourceLoc;
};

export type NewExprNode = {
  kind: "NewExpr";
  typeName?: string;
  typeSpec?: TypeNode;
  args: ExprNode[];
  loc: SourceLoc;
};

export type EOFExprNode = {
  kind: "EOFExpr";
  fileName: string;
  loc: SourceLoc;
};

export type NullExprNode = {
  kind: "NullExpr";
  loc: SourceLoc;
};

export type LValueNode =
  | NameExprNode
  | IndexExprNode
  | FieldExprNode
  | DerefExprNode;

export type DerefExprNode = {
  kind: "DerefExpr";
  expr: ExprNode;
  loc: SourceLoc;
};

export type TypeNode =
  | BasicTypeNode
  | ArrayTypeNode
  | RecordTypeNode
  | EnumTypeNode
  | SetTypeNode
  | PointerTypeNode
  | FileTypeNode
  | NamedTypeNode;

export type BasicTypeNode = {
  kind: "BasicType";
  name: "INTEGER" | "REAL" | "BOOLEAN" | "CHAR" | "STRING" | "DATE";
  loc: SourceLoc;
};

export type ArrayTypeNode = {
  kind: "ArrayType";
  bounds: ArrayBoundNode[];
  elementType: TypeNode;
  loc: SourceLoc;
};

export type ArrayBoundNode = {
  low: number;
  high: number;
  loc: SourceLoc;
};

export type RecordTypeNode = {
  kind: "RecordType";
  fields: VarDeclNode[];
  loc: SourceLoc;
};

export type EnumTypeNode = {
  kind: "EnumType";
  members: string[];
  loc: SourceLoc;
};

export type SetTypeNode = {
  kind: "SetType";
  baseName: string;
  loc: SourceLoc;
};

export type PointerTypeNode = {
  kind: "PointerType";
  target: TypeNode;
  loc: SourceLoc;
};

export type FileTypeNode =
  | { kind: "TextFileType"; loc: SourceLoc }
  | { kind: "RandomFileType"; recordName: string; loc: SourceLoc };

export type NamedTypeNode = {
  kind: "NamedType";
  name: string;
  loc: SourceLoc;
};
